internal/metadata/store.go

```diff
@@ -27,40 +37,46 @@ func (c *ChainMetadataStore) Start() {
 }
 
 func (c *ChainMetadataStore) updateHeightForGroup(groupID string, currentBlockHeight uint64) {
-	c.maxHeightByGroupID[groupID] = max(c.maxHeightByGroupID[groupID], currentBlockHeight)
+	c.maxHeightByGroupID[groupID] = lo.Max([]uint64{c.maxHeightByGroupID[groupID], currentBlockHeight})
 }
 
-func max(a, b uint64) uint64 {
-	if a > b {
-		return a
-	}
+func (c *ChainMetadataStore) updateHeightForUpstream(groupID, upstreamID string, blockHeight uint64) {
+	c.heightByUpstreamID[groupID+upstreamID] = blockHeight
+}
 
-	return b
+func (c *ChainMetadataStore) updateErrorForUpstream(groupID, upstreamID string, err error) {
+	c.errorByUpstreamID[groupID+upstreamID] = err
 }
 
-func (c *ChainMetadataStore) GetGlobalMaxHeight() uint64 {
-	returnChannel := make(chan uint64)
+func (c *ChainMetadataStore) GetBlockHeightStatus(groupID, upstreamID string) BlockHeightStatus {
+	returnChannel := make(chan BlockHeightStatus)
 	c.opChannel <- func() {
-		returnChannel <- c.globalMaxHeight
+		blockHeightStatus := BlockHeightStatus{
+			Error:                c.errorByUpstreamID[groupID+upstreamID],
```

Comment:
Good looks on this.

In other parts of the code, we assume that upstreamID is unique, [for example here](https://github.com/satsuma-data/node-gateway/blob/e6e71622712e4563e91a66ff42986b2b00a3891b/internal/checks/manager.go#L82). 

Although currently, we don't fail out if we define duplicate upstreamIDs in the config. So maybe this should be a check when validating the config. We validate for uniqueness for [groups](https://github.com/satsuma-data/node-gateway/blob/de5533b84efa6b1f906479756f44f3393a9b96b6/internal/config/config.go#L122-L142), so I think we can do it for upstreamIDs (within a single chain config).  Maybe [here](https://github.com/satsuma-data/node-gateway/blob/de5533b84efa6b1f906479756f44f3393a9b96b6/internal/config/config.go#L162)? WDYT

Reply:
Thanks for the links!  I ended up changing `GetUpstreamStatus` to use `groupID+upstreamID` as well.  Also added in validation in the places you linked for checking that upstream IDs are unique within a single group.

Reply:
My bad, I think I was unclear - do we need to make the key "group ID + upstream ID"? Wouldn't upstreamID be enough?

Each _chain_ has its [own set of components](https://github.com/satsuma-data/node-gateway/blob/e0564fadc5bfde876c9b53977af40e94e2728290/internal/server/object_graph.go#L29)  (healthcheck manager, metadata store, etc). 

After the validation you added at the `config.go` level , there shouldn't be any overlap of upstream IDs.